File: iOS/Pool/Sources/Pool/ManageGamblers/ManageGamblerItem.swift
import SwiftUI
import UI
import Account
import ViewingState

struct ManageGamblerItem: View {
    let state: MutationState<PoolMemberModel>

    @Environment(\.boxSpacing) private var boxSpacing

    private var member: PoolMemberModel { state.activeValue() }
    private var isDeleting: Bool { state.isMutating() || state.isMutated() }

    var body: some View {
        HStack(spacing: boxSpacing.medium) {
            avatar
                .frame(width: avatarSize, height: avatarSize)
                .clipShape(Circle())

            VStack(alignment: .leading, spacing: boxSpacing.small) {
                Text(member.gamblerUsername)
                    .fontWeight(.medium)
                    .foregroundStyle(member.isOwner ? Color(sharedResource: .onPrimaryContainter) : Color.primary)
                    .lineLimit(1)
                    .truncationMode(.tail)

                Text(member.gamblerEmail)
                    .font(.subheadline)
                    .foregroundStyle(member.isOwner ? Color(sharedResource: .onPrimaryContainter).opacity(0.7) : Color.secondary)
                    .lineLimit(1)
                    .truncationMode(.tail)
            }

            Spacer(minLength: 0)
        }
        .opacity(isDeleting ? 0.55 : 1)
        .animation(.default, value: isDeleting)
    }

    @ViewBuilder
    private var avatar: some View {
        if isDeleting {
            ZStack {
                Circle().fill(Color(sharedResource: .surfaceVariant))
                BallSpinner()
                    .frame(width: avatarSize * 0.6, height: avatarSize * 0.6)
            }
            .transition(.opacity)
        } else {
            EmailAvatar(email: member.gamblerEmail)
                .transition(.opacity)
        }
    }
}

private let avatarSize: CGFloat = 40

#Preview("Default") {
    ManageGamblerItem(
        state: .idle(
            PoolMemberModel(
                gamblerId: "1",
                gamblerUsername: "danielsanto",
                gamblerEmail: "daniel@example.com"
            )
        )
    )
    .padding()
}

#Preview("Deleting") {
    ManageGamblerItem(
        state: .mutating(
            PoolMemberModel(
                gamblerId: "1",
                gamblerUsername: "danielsanto",
                gamblerEmail: "daniel@example.com"
            )
        )
    )
    .padding()
}

File: Android/pool/src/main/java/com/felipearpa/tyche/pool/managegamblers/ManageGamblerItem.kt
package com.felipearpa.tyche.pool.managegamblers

import androidx.compose.animation.Crossfade
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.unit.dp
import com.felipearpa.tyche.account.EmailAvatar
import com.felipearpa.tyche.ui.loading.BallSpinner
import com.felipearpa.tyche.ui.theme.LocalBoxSpacing
import com.felipearpa.tyche.ui.theme.TycheTheme
import com.felipearpa.ui.state.MutationState
import com.felipearpa.ui.state.activeValue

@Composable
fun ManageGamblerItem(
    state: MutationState<PoolMemberModel>,
    modifier: Modifier = Modifier,
) {
    val member = state.activeValue()
    val isDeleting = state is MutationState.Mutating || state is MutationState.Mutated

    val rowAlpha by animateFloatAsState(
        targetValue = if (isDeleting) 0.55f else 1f,
        label = "rowAlpha",
    )

    Row(
        modifier = modifier
            .fillMaxWidth()
            .alpha(rowAlpha),
        horizontalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.medium),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Crossfade(targetState = isDeleting, label = "avatar") { deleting ->
            if (deleting) {
                Surface(
                    shape = CircleShape,
                    color = MaterialTheme.colorScheme.surfaceVariant,
                    modifier = Modifier.size(avatarSize),
                ) {
                    Box(contentAlignment = Alignment.Center) {
                        BallSpinner(modifier = Modifier.size(avatarSize * 0.6f))
                    }
                }
            } else {
                EmailAvatar(
                    email = member.gamblerEmail,
                    modifier = Modifier
                        .size(avatarSize)
                        .clip(CircleShape),
                )
            }
        }

        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = member.gamblerUsername,
                style = MaterialTheme.typography.bodyLarge.copy(fontWeight = FontWeight.Medium),
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                text = member.gamblerEmail,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}

private val avatarSize = 40.dp

@PreviewLightDark
@Composable
private fun ManageGamblerItemPreview() {
    TycheTheme {
        Surface {
            ManageGamblerItem(
                state = MutationState.Idle(
                    PoolMemberModel(
                        gamblerId = "1",
                        gamblerUsername = "danielsanto",
                        gamblerEmail = "daniel@example.com",
                    ),
                ),
            )
        }
    }
}

@PreviewLightDark
@Composable
private fun ManageGamblerItemDeletingPreview() {
    val member = PoolMemberModel(
        gamblerId = "1",
        gamblerUsername = "danielsanto",
        gamblerEmail = "daniel@example.com",
    )
    TycheTheme {
        Surface {
            ManageGamblerItem(
                state = MutationState.Mutating(original = member, updated = member),
            )
        }
    }
}
